src/posts/abstracting-the-frontend/index.md

```diff
@@ -0,0 +1,124 @@
+
+1. The core of a design system should be reusable across products. A company's products don't have to look the same,
+   but they should feel the same.
+2. Component libraries should exist that strictly follow the rules of the design system.
+3. A component library should not consume remote data directly.
+4. A design system's documentation should be available to the public.
+
+**Note:** Point 3 makes the layer approach a bit awkward, but I hope we can agree on the general philosophy.
+
+#### Data aggregate
+
+If we ever assumed that the backend team is only concerned delivering data to our front ends, this is no longer the case.
+Backend developers building API's have their own group of stakeholders to worry about, integrators. This means that there will be
+certain compromises in API design. Thus, having a middleware where we can create new datasets tailored
+to the front end can prove quite useful.
+
+A front end that does not communicate with a data source is a rare sight these days and I believe that the number of data points a
+single client side application will have to interact with will only grow. A single resource in our front end might consist of
+multiple different sources of data, so in order to reduce complexity in our application code, we will want to host an aggregate of the
+data our front ends interact with so that we can simplify what our client will consume.
+
+#### The application
+
+This one might not come as a big surprise. This is the layer that separates a fridge from a smart watch.
+But provided that we have the above layers, maintaining our applications should now be a less daunting task.
+I had thought about splitting this up into a logic and view layer, but being a firm believer in colocation I feel that
+sharing logic between applications is not something that needs to be asserted in order to succeed.
+There are people out there far smarter than me that will deep dive into what makes a scalable front-end application.
+
+Check out Kent C. Dodds' post on [Colocation](https://kentcdodds.com/blog/colocation) if you're interested in learning more about the principle.
+
+### Practical examples
+
+Contrary to my introductory statement, that this post is meant to provide a higher level overview of the types of lines we should be
+drawing in our front-end architecture, I would like to shed a light on some concrete options that I would see myself reaching for to
+follow this pattern.
+
+#### Storybook
+
+[Storybook](https://storybook.js.org/) is a great option for creating design systems, as they combine the elements of documentation
+and component library in a very nice way. Storybook is framework agnostic but provides extensions for most of the more popular
+libraries for building components. Whatever framework you choose to build your component library in probably comes down to where
+your product makes the most impact.
+
+In terms of the actual design system, this does not rely on a product or a technology.
```

Comment:
Maybe it's kind of a litmus test for any given component/tool of your design system stack could be - "could I replace this concept/tool/library out without changing any of the other components?"

Reply:
This would then mean that "The one thing that you couldn't replace, that IS your design system."